src/OpenTelemetry/Trace/ActivitySampler.cs

```diff
@@ -0,0 +1,48 @@
+﻿// <copyright file="ActivitySampler.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace OpenTelemetry.Trace
+{
+    /// <summary>
+    /// Sampler to reduce data volume. This sampler executes before Activity object is created.
+    /// </summary>
+    public abstract class ActivitySampler
+    {
+        /// <summary>
+        /// Gets the sampler description.
+        /// </summary>
+        public abstract string Description { get; }
+
+        /// <summary>
+        /// Checks whether activity needs to be created and tracked.
+        /// </summary>
+        /// <param name="parentContext">Parent activity context. Typically taken from the wire.</param>
+        /// <param name="traceId">Trace ID of a activity to be created.</param>
+        /// <param name="spanId">Span ID of a activity to be created.</param>
```

Comment:
I wonder whats the value of spanid here - the span/activity is not yet created, and will be done based on the outcome of this method. 
Can't think of the use of spanid in a typical sampling algorithm, as new spanid is just random.
(With activity there is no option to know the spanid ahead of Activity creation, and no option to change spanid after creation).

Reply:
IIRC an OTel SDK should be able to generate the IDs before calling the sampler, and historically some distributed tracing systems actually used the span ID instead of the trace ID to make their decision.

Reply:
If this is a requirement, we need to address this,
Adding it here for tracking: https://github.com/open-telemetry/opentelemetry-dotnet/issues/675#issuecomment-631080503

cc: @tarekgh

Reply:
> IIRC an OTel SDK should be able to generate the IDs before calling the sampler, and historically some distributed tracing systems actually used the span ID instead of the trace ID to make their decision.

I am wondering how a SpanId (for a span which not created yet) will be used in the sampling decision? this looks weird to me as this Id never been used till the sampling time. do we expect the samplers recognize specific patterns of Span Ids to make the decision accordingly? I can understand it could make sense to use the Span Id for sampling if the span is already created but not if the span really not created at all.

@cijothomas maybe the trace id is interesting too here.

@noahfalk

Reply:
since traceid for parent, and the child-to-be-created are going to be the same, I dont see how traceid is an issue? 
Unless parent is null?

Reply:
> I can understand it could make sense to use the Span Id for sampling if the span is already created but not if the span really not created at all.

@tarekgh in practice the user code will only see the span after the sampling decision was already taken, remember that on OTel a span or a placeholder for it is always returned anyway so making the decision on the SpanId is fine. In the case that I vaguely remember it was probabilistic samples of spans instead of traces but that is something that most users consider undesirable: fragments of traces are not useful unless you have a really large amount of data. I think that nowadays the standard is to sample traces not spans.

Reply:
@cijothomas

> since traceid for parent, and the child-to-be-created are going to be the same, I dont see how traceid is an issue?

here is what the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md) says:
`TraceId of the Span to be created. It can be different from the TraceId in the SpanContext. Typically in situations when the Span to be created starts a new Trace.`

which means the trace id can be different than the parent one. right?

Reply:
@pjanotti thanks for the clarification. .NET doesn't have the notion of `Span placeholder`. .NET now supports the way to do the sampling decision before creating the Activity/Span to avoid allocating objects which not going to be used if it is sampled out. so SDK implementation can either decide to create the activity anyway before sampling (which will create the span and trace id for it) and then use it in sampling, or just ignore the span id part in the sampling decision and avoid creating Activity till it is really needed. 

CC @reyang

Reply:
Samplers decide whether to create a span or not, a sampler will not have access to the `SpanId`. Samplers will have access to:
1. All arguments that are passed to [span creation API](https://github.com/open-telemetry/opentelemetry-specification/blob/f9aabe2cb1cd6b329cf61375447d97a87113a84f/specification/trace/api.md#span-creation).
2. Context - as samplers run within the same execution context (e.g. thread / async task).

Here goes the Python sampler for reference: https://github.com/open-telemetry/opentelemetry-python/blob/abdfa12dfd76613bef5cc5e9266439e983074bbc/opentelemetry-api/src/opentelemetry/trace/sampling.py

Reply:
I doubt that any default OTel SDK is shipping anything sampling based on SpanId. The probabilistic sampler should be based on TraceId. 

Per the discussion above I'm not sure if we are okay regarding the probabilistic sampler. Could you please confirm that? @tarekgh @cijothomas 

Let me see if I can dig some more info about this.

Reply:
@reyang, what you mentioned makes sense to me and that was my understanding. the question now, does the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md) need to change then?

@pjanotti 
> Per the discussion above I'm not sure if we are okay regarding the probabilistic sampler. Could you please confirm that?

I am not in favor of having the span id part of the sampling decision. as @reyang mentioned, the sampling decision should be based on the same parameters used for span creation. Maybe the spec need to get updated?

Reply:
The probability sampler being implemented is following https://github.com/open-telemetry/opentelemetry-specification/pull/331, and now @c24t is trying to make it part of the spec.

* Consider implementations in other projects:
  - [Java](https://github.com/open-telemetry/opentelemetry-java/blob/master/sdk/src/main/java/io/opentelemetry/sdk/trace/Sampler.java)
  - [Go](https://github.com/open-telemetry/opentelemetry-go/blob/master/api/trace/sampler.go)
  - [Python](https://github.com/open-telemetry/opentelemetry-python/blob/master/opentelemetry-api/src/opentelemetry/trace/sampling.py)

Reply:
@reyang I am bit confused with this:
From this line, it looks like Python sampler also takes span_id as input to make sampling decision.
https://github.com/open-telemetry/opentelemetry-python/blob/abdfa12dfd76613bef5cc5e9266439e983074bbc/opentelemetry-api/src/opentelemetry/trace/sampling.py#L72

This also matches spec, which states sampling takes "spanid of the span to be created"
https://github.com/open-telemetry/opentelemetry-specification/blob/f9aabe2cb1cd6b329cf61375447d97a87113a84f/specification/trace/sdk.md#shouldsample

Reply:
> @reyang I am bit confused with this:
> From this line, it looks like Python sampler also takes span_id as input to make sampling decision.
> https://github.com/open-telemetry/opentelemetry-python/blob/abdfa12dfd76613bef5cc5e9266439e983074bbc/opentelemetry-api/src/opentelemetry/trace/sampling.py#L72

It seems that span id is actually the "parent id" :)
https://github.com/open-telemetry/opentelemetry-python/blob/4ab3978cf71af6425a120e9ebfb9960b27987b7a/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L655

Reply:
then why we have spanId and traceId parameters at all then? it looks redundant. Also, the spec says the trace Id and trace Id are the values that the Span is going to be created with.

Reply:
> @cijothomas
> 
> > since traceid for parent, and the child-to-be-created are going to be the same, I dont see how traceid is an issue?
> 
> here is what the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md) says:
> `TraceId of the Span to be created. It can be different from the TraceId in the SpanContext. Typically in situations when the Span to be created starts a new Trace.`
> 
> which means the trace id can be different than the parent one. right?

Yes.
mostly when parent is null, as we generate new traceid.

Reply:
See if we can get some insights from our Python/sampling expert.

@c24t wish to get your insights on this.
* The [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/f9aabe2cb1cd6b329cf61375447d97a87113a84f/specification/trace/sdk.md#shouldsample) is saying that samplers should have access to "SpanId of the Span to be created", which seems strange? (why would we want to generate a random number and discard it if not sampled).
* The [Python implementation](https://github.com/open-telemetry/opentelemetry-python/blob/4ab3978cf71af6425a120e9ebfb9960b27987b7a/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L655) seems to align with my understanding - we take the span id from context (the id of the parent span). And if the sampler decided to NOT_SAMPLE, no new span id will be created (and a dummy span will be returned with [0 as the span id](https://github.com/open-telemetry/opentelemetry-python/blob/090b6640495bcb994a9467ab23e2b1abe7fc4af9/opentelemetry-api/src/opentelemetry/trace/__init__.py#L439)).

Reply:
@cijothomas

> mostly when parent is null, as we generate new traceid.

I am wondering how this will be useful to the sampler? if the parent is null, then the sampler will know that and can decide without any need for spanid, right?

Reply:
@reyang the context.span_id in Python carries a new SpanId: https://github.com/open-telemetry/opentelemetry-python/blob/4ab3978cf71af6425a120e9ebfb9960b27987b7a/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L642

@tarekgh 
> I am wondering how this will be useful to the sampler? if the parent is null, then the sampler will know that and can decide without any need for spanid, right? 

The SpanId seems a legacy that we can perhaps remove from the spec (we may find someone that really wants to keep it). The TraceId I don't see a way around it: if there is no parent span we need to generate a random thing (in this case the TraceId) to decide if it is going to be sampled or not.

Reply:
@pjanotti 

> The TraceId I don't see a way around it: if there is no parent span we need to generate a random thing (in this case the TraceId) to decide if it is going to be sampled or not.

if there is no parent, then you can create a new SpanContext, fill it with whatever info you want, and then pass it to the sampler (as the parent). would that be wrong? And if the sampler decides to create the span (or sample it in), the created SpanContext can still be passed to the Span creation as a parent. I think that what python already doing it?

Reply:
@tarekgh if we can do the same as Python we are good. So it seems that on the probabilistic sampler we write that code.

Reply:
I believe with the current offered APIs we can do as what python is doing. @cijothomas @reyang  what you think?

Reply:
**Capturing my understanding (to make us in sync with spec/python)**
Always generate new spanid
Generate new traceid if parent is null. Else use traceid of parent.
Feed these to the Sampler algorithm.
If sampled=yes
{
create activity based on the traceid,spanid from above.
}
else
{
dont create activity. the price paid to generate traceid/spanid is wasted, but we are okay with that.
}

The above should get us match python/spec.

Additionaly, we can follow up with spec to see if we can modify Sampler interface should match the Span creation parameters + traceid as well.

Let me know if my understanding is correct.

Reply:
We could do some optimization based on the type of the sampler since always-on/off ignore all that info but that is just an optimization :)

Reply:
> I believe with the current offered APIs we can do as what python is doing. @cijothomas @reyang what you think?

Yes. This can be done in SDK without requiring changes in Activity APIs.

Reply:
> create activity based on the traceid,spanid from above.

to be made clearer, these TraceId and SpanId will passed to the Activity creation as parent ActivityContext. in another word, the created activity can end up with a new Span Id but will inherit the TraceId from the parent.

Reply:
@tarekgh thanks for correcting me - Since the SpanID will be ignored anyway by the Activity, then we'll not have python like solution, as in python the generated spanid will be used. But for activity the generated spanid will be ignored.

This is the spanid which was given to sampler - if sampler (custom one) makes decision based on spanid, then we **must** use it when Activity is created. 

But this brings back to the spec question - why would sampler need SpanId ?

Reply:
@cijothomas 
> why would sampler need SpanId ?

This is for samplers doing statistical sampling of spans not traces - I personally don't see much value on it since it results in broken traces. That said we should pursue a change in the specs if we don't want to support it at all.

Reply:
@reyang can we track this to be discussed in the OT specs forum?

Reply:
I need to dig a bit into the history.

Reply:
Okay, here goes the spec PR which added `SpanID` to sampler https://github.com/open-telemetry/opentelemetry-specification/pull/81. We will need to learn the history from @SergeyKanzhelev.

Reply:
@reyang here and there I tried to dig something about this and I didn't see any PR/discussion mentioning the `SpanID` perhaps it was an oversight? Unless @SergeyKanzhelev has something to add we should propose a simple change of removing it to see if anyone complains :)

Reply:
So far my archeology work traced back to [OpenCensus](https://github.com/census-instrumentation/opencensus-python/blob/99821a0b826445cecce0881a01790c8cc6958ceb/opencensus/trace/samplers/__init__.py#L25) 😆

Reply:
yeah 🤠 (they didn't have an Indiana Jones emoji but the cowboy has a hat 🤷‍♂️), it seems very much like OpenCensus - that said everything that I see is using traceId to make their decisions, now will be the moment to remove it before some "creative person" starts to depend on it :)

Reply:
I've added the topic to **Specification SIG meeting 05/26/2020** agenda.

Reply:
Discussed during the SIG meeting and agreed to remove it from the spec.

https://github.com/open-telemetry/opentelemetry-specification/pull/621